web/server/net_logger.go

```diff
@@ -328,22 +331,27 @@ func (w *remoteLogWriterGRPC) write(logs []*apppb.LogEntry) error {
 }
 
 func (w *remoteLogWriterGRPC) getOrCreateClient(ctx context.Context) (apppb.RobotServiceClient, error) {
-	if w.service == nil {
-		w.clientMutex.Lock()
-		defer w.clientMutex.Unlock()
+	w.clientMutex.Lock()
+	defer w.clientMutex.Unlock()
 
-		client, err := config.CreateNewGRPCClient(ctx, w.cfg, w.loggerWithoutNet)
-		if err != nil {
-			return nil, err
-		}
+	if w.service != nil {
+		return w.service, nil
+	}
 
-		w.rpcClient = client
-		w.service = apppb.NewRobotServiceClient(w.rpcClient)
+	client, err := config.CreateNewGRPCClient(ctx, w.cfg, w.loggerWithoutNet)
+	if err != nil {
+		return nil, err
 	}
+
+	w.rpcClient = client
+	w.service = apppb.NewRobotServiceClient(w.rpcClient)
 	return w.service, nil
 }
 
 func (w *remoteLogWriterGRPC) close() {
+	w.clientMutex.Lock()
+	defer w.clientMutex.Unlock()
+
 	if w.rpcClient != nil {
```

Comment:
My reading of the code suggests this is a perfectly synchronized read of the `w.rpcClient` member that won't race with a write. But one can't come to that conclusion by only reading the `remoteLogWriter` implementation. So I thought it best to make the implementation obviously correct from a data race perspective and resilient to other code changing which makes this incorrect.

Certainly if close is racing with writers trying to send over the client -- we're in a bad place anyways. But similarly I don't see a need to poke the data race bear.

Reply:
I think this lock is wise to put in despite the attempt to clear the log queue in `Close`.